room/compiler/src/main/kotlin/androidx/room/processor/CustomConverterProcessor.kt

```diff
@@ -77,29 +78,38 @@ class CustomConverterProcessor(val context: Context, val element: XTypeElement)
         val converterMethods = methods.filter {
             it.hasAnnotation(TypeConverter::class)
         }
+        val typeConverterFactory = declaredType.asTypeElement()
+            .toAnnotationBox(TypeConverterFactory::class)
+            ?.getAsType("value")
         context.checker.check(converterMethods.isNotEmpty(), element, TYPE_CONVERTER_EMPTY_CLASS)
         val allStatic = converterMethods.all { it.isStatic() }
         val constructors = element.getConstructors()
         val isKotlinObjectDeclaration = element.isKotlinObject()
-        context.checker.check(
-            isKotlinObjectDeclaration || allStatic || constructors.isEmpty() ||
-                    constructors.any {
-                        it.parameters.isEmpty()
-                    }, element, TYPE_CONVERTER_MISSING_NOARG_CONSTRUCTOR
-        )
+        if (typeConverterFactory == null) {
+            context.checker.check(
+                isKotlinObjectDeclaration || allStatic || constructors.isEmpty() ||
+                        constructors.any {
+                            it.parameters.isEmpty()
+                        }, element, TYPE_CONVERTER_MISSING_NOARG_CONSTRUCTOR
```

Comment:
we probably want to update this message (in processor errors) to refer to the type converter factory if they can provide the instance.